storage/src/binary/mod.rs

```diff
@@ -0,0 +1,457 @@
+//! A stateless Binary Merkle Tree.
+//!
+//! The tree is constructed level-by-level. Level 0 consists of the leaf nodes. At each higher
+//! level, pairs of nodes are combined using the provided hasher. If a level contains an odd number
+//! of nodes, the missing sibling is replaced with `H::Digest::default()`.
+//!
+//! For example, given three leaves A, B, and C, the tree is constructed as follows:
+//!
+//! ```text
+//!     Level 2 (root):       [combine(combine(A,B),combine(C,DEFAULT))]
+//!     Level 1:              [combine(A,B),combine(C,DEFAULT)]
+//!     Level 0 (leaves):     [A,B,C]
+//! ```
+//!
+//! A Merkle proof for a given leaf is generated by collecting the sibling at each level (from the leaf
+//! up to the root). The proof can then be used to verify that the leaf is part of the tree.
+//!
+//! This data structure is often used to generate a root for a block digest (over included transactions).
+//!
+//! # Example
+//!
+//! ```rust
+//! use commonware_storage::binary::Tree;
+//! use commonware_cryptography::{hash, Sha256, sha256::Digest};
+//!
+//! // Build tree
+//! let txs = [b"tx1", b"tx2", b"tx3", b"tx4"];
+//! let digests: Vec<Digest> = txs.iter().map(|tx| hash(*tx)).collect();
+//! let mut hasher = Sha256::default();
+//! let tree = Tree::new(&mut hasher, digests.clone()).unwrap();
+//! let root = tree.root();
+//!
+//! // Generate a proof for leaf at index 1.
+//! let proof = tree.prove(0).unwrap();
+//! assert!(proof.verify(&mut hasher, &digests[0], 0, &root));
+//! ```
+
+use commonware_cryptography::{Digest, Hasher};
+use std::mem::size_of;
+
+/// Combines two digests into a new digest.
+fn combine<H: Hasher>(hasher: &mut H, left: &H::Digest, right: &H::Digest) -> H::Digest {
+    hasher.update(left);
+    hasher.update(right);
+    hasher.finalize()
+}
+
+/// A stateless Binary Merkle Tree that computes a root over an arbitrary set
+/// of [commonware_cryptography::Digest].
+#[derive(Clone, Debug)]
+pub struct Tree<H: Hasher> {
+    /// Levels of the tree: level 0 contains the leaves, and each subsequent level
+    /// contains the parent nodes computed from the previous level.
+    ///
+    /// The last level contains a single node: the root.
+    levels: Vec<Vec<H::Digest>>,
+}
+
+impl<H: Hasher> Tree<H> {
+    /// Builds a Merkle Tree from a slice of leaf digests.
+    ///
+    /// If `leaves` is empty, returns `None`.
+    ///
+    /// # Warning
+    ///
+    /// It is not safe to insert `H::Digest::default()` as a leaf. This could lead to
+    /// a proof being generated for a non-existent leaf.
```

Comment:
Maybe make this panic or return error if someone tries to insert a default() hash?

Reply:
No longer possible to create a `default` Digest, let alone add it 😉